Attempt to fix grapples usable by firing before punching flag

ClobberXD opened the pull request:
Check for player carrying flag in hook entity's `on_step` as well. Fixes #600, by eliminating the `TOC ~= TOU` issue that's responsible for the bug.

If a player takes the flag while their hook is in the air, the hook is returned to the player and the following chat message is displayed:

`You can't use grapples whilst carrying the flag`

Ready for review.

### Testing instructions

- Go close to the enemy flag.
- Shoot the grapple far away and punch the flag before the grapple lands.
- Observe that the grapple is returned, with the aforementioned message in chat.
- Also verify that the grapple works as expected when you don't have the flag.

ClobberXD commented:
Tested; works. PR is ready for review.

Thomas--S commented:
If you'd prefer to not duplicate the code, something like this could work (pseudo code):

```lua
old_on_step = registered_entities["hook"].on_step
registered_entities["hook"].on_step = function ()
    if (all kind of funny CTF checks) then
        return false
    end
    return old_on_step()
end
```

ClobberXD commented:
I for some reason was thinking that wouldn't work. What an idiot I am, of course that's possible and the better option. :)

Thomas--S commented on mods/ctf/ctf_classes/ranged.lua:
```diff
@@ -69,3 +69,21 @@ end
+		self.object:remove()
+		return
+	end
+	return old_grapple_step
```
Are you sure that this works?

I think there should be paranthesis, like this: `return old_grapple_step(self, dtime)`.

But it is possible that I'm wrong here, I didn't test it.

LoneWolfHT replied:
It should be `old_grapple_step(self, dtime, ...)` and you'll need to add the `...` to line 75 too.
The '...' in this case is all the potential params after `dtime`

ClobberXD replied:
That's rather stupid, who even wrote this dumb code?

ClobberXD replied:
To whoever wrote this stupid code: I've fixed it for you.

ClobberXD commented:
Fixed `old_grapple_step` not being invoked properly, and fixed the function def so that it overrides in a backwards-compatible manner.

Works as expected. I've re-tested using grapples and grapple guns, both with and without the flag.

LoneWolfHT commented:
Do we want to convert this to use an on_collide while we're at it? (via grenades mod I guess)

ClobberXD commented:
> Do we want to convert this to use an on_collide while we're at it? (via grenades mod I guess)

I'd prefer to not do that right now, as we have v2.0.0 to think about. The objective of this PR is to fix the ability for players to abuse grapples, and that objective has been reached.

While I like the idea of using the grenades API for this, it's a complete overhaul of grapples, and as such, I'd like to see that happen after v2.0.
